### ISSUE TITLE
es6jsx: no-unused-vars misses JSX references

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#1534]
I'm using eslint v0.11.0-alpha.0 and getting `no-unused-vars` errors in my JSX template code:

```JavaScript
var React = require('react');

var App = React.createClass({
  render: function() {
    return (
      <div>
        <h1>wat</h1>
      </div>
    );
  }
});

React.render(<App />, document.body);
```

This produces the following errors:

```bash
ಠ_ಠ  ~/code/happytrack λ eslint durp.jsx

durp.jsx
  3:4   error  App is defined but never used                                 no-unused-vars
```

**Comment by yannickcr** [comment]
Hi,

It seems we still miss some JSX references.

I still got the `App is defined but never used` error with the above example with eslint master branch.

**Comment by lo1tuma** [comment]
I can confirm this, the following test case fails:

```js
{ code: "var App; var bar = React.render(<App/>); console.log(bar);", args: [1, {vars: "all"}], ecmaFeatures: { jsx: true } },
```

**Comment by nzakas** [comment]
@yannickcr are you using the master branch?

**Comment by yannickcr** [comment]
@nzakas Yes, I'm currently testing the master branch 750128262fc322fd2c143a0b3440a7080829c4d2 to validate our React components.

**Comment by lo1tuma** [comment]
Can  this be reopened? It still does not work if you have a global JSX variable but use it in a function scope.

Example:
```js
var App;

function foo() {
    React.render(<App />);
}

foo();
```

Output:
> App is defined but never used

**Comment by MoOx** [comment]
5f2ee66101b21d005bcb55bc73b3030eff95c8e3 seems to have a similar test.

**Comment by lo1tuma** [comment]
@MoOx No, the test doesn’t have an additional function scope as in my example above.

**Comment by pstadler** [comment]
I think there's still a bug related with jsx and no-unused-vars in 0.17.1

**jsx-failing.js**
```javascript
var React = require('react')
  , Navigation = require('./navigation')

module.exports = React.createClass({
  render: function() {
    return (
      <div>
        <Navigation />
        <div className="container">
          {this.props.children}
        </div>
      </div>
    )
  }
})
```

**.eslintrc:**
```json
{
  "env": {
    "node": true
  , "browser": true
  },
  "ecmaFeatures": {
    "jsx": true
  },
  "rules": {
    "semi": false,
    "quotes": "single",
    "comma-style": [2, "first"],
    "strict": false,
    "eol-last": false,
    "no-mixed-requires": false
  }
}
```

**Result:**
```sh
$ eslint --version
v0.17.1

$ eslint --config .eslintrc jsx-failing.js

jsx-failing.js
  2:4  error  Navigation is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by MoOx** [comment]
@pstadler you need to use eslint-plugin-react to handle this.

**Comment by pjjanak** [comment]
I'm also still seeing the issue mentioned. I'm fairly certain this used to work before I upgraded to the most current version of eslint. Since this is part of the JSX spec and not specifically related to React, I would think this would work...I don't expect to have to use a plugin for something that's presumably all ready supported by eslint.

**Comment by pjjanak** [comment]
Ah...nevermind. Just saw this blog post http://eslint.org/blog/2015/03/eslint-0.17.0-released/

**Comment by pstadler** [comment]
Ok, that makes sense. Thanks for the feedback.
